x-pack/plugin/ml/qa/native-multi-node-tests/src/test/java/org/elasticsearch/xpack/ml/integration/ClassificationIT.java

```diff
@@ -497,6 +501,59 @@ public void testTwoJobsWithSameRandomizeSeedUseSameTrainingSet() throws Exceptio
         assertThat(secondRunTrainingRowsIds, equalTo(firstRunTrainingRowsIds));
     }
 
+    public void testSetUpgradeMode_ExistingTaskGetsUnassigned() throws Exception {
+        initialize("classification_set_upgrade_mode");
+        indexData(sourceIndex, 300, 0, KEYWORD_FIELD);
+
+        assertThat(upgradeMode(), is(false));
+
+        DataFrameAnalyticsConfig config = buildAnalytics(jobId, sourceIndex, destIndex, null, new Classification(KEYWORD_FIELD));
+        registerAnalytics(config);
+        putAnalytics(config);
+        startAnalytics(jobId);
+        assertThat(analyticsTaskList(), hasSize(1));
+        assertThat(analyticsAssignedTaskList(), hasSize(1));
+
+        setUpgradeModeTo(true);
+        assertThat(analyticsTaskList(), hasSize(1));
+        assertThat(analyticsAssignedTaskList(), is(empty()));
+
+        GetDataFrameAnalyticsStatsAction.Response.Stats analyticsStats = getAnalyticsStats(jobId);
+        assertThat(analyticsStats.getState(), is(equalTo(DataFrameAnalyticsState.STARTING)));
+        assertThat(analyticsStats.getAssignmentExplanation(), is(equalTo(AWAITING_UPGRADE.getExplanation())));
+        assertThat(analyticsStats.getNode(), is(nullValue()));
+
+        setUpgradeModeTo(false);
+        assertThat(analyticsTaskList(), hasSize(1));
+        assertThat(analyticsAssignedTaskList(), hasSize(1));
+
+        analyticsStats = getAnalyticsStats(jobId);
+        assertThat(analyticsStats.getState(), is(equalTo(DataFrameAnalyticsState.STARTING)));
```

Comment:
Isn't this assertion risky? Could it not be the task has advanced state by this point?

Reply:
You're right. I've removed this assertion and added `waitUntilAnalyticsIsStopped` instead.
I'm wondering though what to do with an assertion in line 522. Do you think it is reasonable to expect it to be `STARTING`?

Reply:
I thought that one made sense because the persistent task exists but it won't be assigned as upgrade mode was enabled. This the reported state should consistently be `starting`. Am I not reading this right?

Reply:
I'm wondering if the state could advance between lines 513 (starting job) and 517. (pausing it with upgrade mode)

Reply:
Perhaps you're right. You could manually test this by running a job and setting upgrade mode while it's past the `starting` state.

Reply:
I've just removed that state assertion too. Job state is not that important to verify in this test.
Upgrade mode should work correctly for any job state, so I think there is no need asserting there.